api/src/main/scala/Geoprocessing.scala

```diff
@@ -18,10 +18,124 @@ trait Geoprocessing extends Utils {
+      else rasterGroupedAverage(rasterLayers, targetLayer, aoi)
+
+    ResultDouble(average)
+  }
+
+  /**
+    * Return the average pixel value from a target raster and a MultiPolygon
+    * area of interest.
+    *
+    * @param   targetLayer   The target TileLayerCollection
+    * @param   multiPolygon  The AOI as a MultiPolygon
+    * @return                A one element map averaging the pixel values
+    */
+  private def rasterAverage(
+    targetLayer: TileLayerCollection[SpatialKey],
+    multiPolygon: MultiPolygon
+  ): Map[String, Double] = {
+    val update = (newValue: Double, pixelValue: (DoubleAdder, LongAdder)) => {
+      pixelValue match {
+        case (accumulator, counter) => accumulator.add(newValue); counter.increment()
+      }
+    }
+
+    val metadata = targetLayer.metadata
+    val pixelValue = ( new DoubleAdder, new LongAdder )
+
+    targetLayer.par.foreach({ case (key, tile) =>
+      val re = RasterExtent(metadata.mapTransform(key), metadata.layout.tileCols,
+        metadata.layout.tileRows)
+
+      Rasterizer.foreachCellByMultiPolygon(multiPolygon, re) { case (col, row) =>
+        val targetLayerData = tile.getDouble(col, row)
+
+        val targetLayerValue =
+          if (isData(targetLayerData)) targetLayerData
+          else 0.0
+
+        update(targetLayerValue, pixelValue)
+      }
+    })
+
+    pixelValue match {
+      case (accumulator, counter) => Map("List(0)" -> accumulator.sum / counter.sum)
+    }
+  }
+
+  /**
+    * Return the average pixel value from a target raster and a MultiPolygon
+    * area of interest.
+    *
+    * @param   rasterLayers  A sequence of TileLayerCollections
+    * @param   targetLayer   The target TileLayerCollection
+    * @param   multiPolygon  The AOI as a MultiPolygon
+    * @return                A map of targetRaster pixel value averages
+    */
+  private def rasterGroupedAverage(
+    rasterLayers: Seq[TileLayerCollection[SpatialKey]],
+    targetLayer: TileLayerCollection[SpatialKey],
+    multiPolygon: MultiPolygon
+  ): Map[String, Double] = {
+    val init = () => ( new DoubleAdder, new LongAdder )
+    val update = (newValue: Double, pixelValue: (DoubleAdder, LongAdder)) => {
+      pixelValue match {
+        case (accumulator, counter) => accumulator.add(newValue); counter.increment()
+      }
+    }
+
+    val metadata = targetLayer.metadata
+    var pixelGroups: TrieMap[List[Int], (DoubleAdder, LongAdder)] = TrieMap.empty
+
+    joinCollectionLayers(targetLayer +: rasterLayers).par
+      .foreach({ case (key, targetTile :: tiles) =>
+        val extent: Extent = metadata.mapTransform(key)
+        val re: RasterExtent = RasterExtent(extent, metadata.layout.tileCols,
+            metadata.layout.tileRows)
+
+        Rasterizer.foreachCellByMultiPolygon(multiPolygon, re) { case (col, row) =>
+          val pixelKey: List[Int] = tiles.map(_.get(col, row)).toList
+          var pixelValues = pixelGroups.getOrElseUpdate(pixelKey, init())
```

Comment:
Good catch -- I'll dump this. We can drop the assignment and just do this at the end instead:

```
update(targetLayerValue, pixelGroups.getOrElseUpdate(pixelKey, init()))
```